xs2a-adapter-aspsp-registry/src/main/java/de/adorsys/xs2a/adapter/registry/LuceneAspspRepository.java

```diff
@@ -0,0 +1,190 @@
+        try (IndexWriter indexWriter = new IndexWriter(directory, indexWriterConfig)) {
+            for (Aspsp aspsp : aspsps) {
+                save(indexWriter, aspsp);
+            }
+        } catch (IOException e) {
+            throw new RegistryIOException(e);
+        }
+    }
+
+    @Override
+    public Optional<Aspsp> findById(String id) {
+        ScoreDoc scoreDoc = parseScoreDoc(id);
+        if (scoreDoc == null) {
+            return Optional.empty();
+        }
+
+        return getDocument(scoreDoc.doc).map(document -> {
+            Aspsp aspsp = new Aspsp();
+            aspsp.setId(id);
+            aspsp.setName(deserialize(document.get(NAME_FIELD_NAME)));
+            aspsp.setUrl(deserialize(document.get(URL_FIELD_NAME)));
+            aspsp.setBic(deserialize(document.get(BIC_FIELD_NAME)));
+            aspsp.setBankCode(deserialize(document.get(BANK_CODE_FIELD_NAME)));
+            aspsp.setAdapterId(deserialize(document.get(ADAPTER_ID_FIELD_NAME)));
+            return aspsp;
+        });
+    }
+
+    private Optional<Document> getDocument(int docId) {
+        try (IndexReader indexReader = DirectoryReader.open(directory)) {
+            if (docId < 0 || docId >= indexReader.maxDoc()) {
+                return Optional.empty();  // index reader throws IllegalArgumentException if docID is out of bounds
+            }
+            Document document = indexReader.document(docId);
+            return Optional.of(document);
+        } catch (IOException e) {
+            throw new RegistryIOException(e);
+        }
+    }
+
+    private String deserialize(String s) {
+        if ("null".equals(s)) {
+            return null;
+        }
+        return s;
+    }
+
+    @Override
+    public List<Aspsp> findByBic(String bic, String after, int size) {
+        Query query = getBicQuery(bic);
+        return find(query, after, size);
+    }
+
+    private Query getBicQuery(String bic) {
+        return new PrefixQuery(new Term(BIC_FIELD_NAME, bic));
+    }
+
+    private List<Aspsp> find(Query query, String after, int size) {
+        try (IndexReader indexReader = DirectoryReader.open(directory)) {
+            IndexSearcher indexSearcher = new IndexSearcher(indexReader);
+            ScoreDoc afterDoc = parseScoreDoc(after);
+            TopDocs topDocs = indexSearcher.searchAfter(afterDoc, query, size);
+            ScoreDoc[] scoreDocs = topDocs.scoreDocs;
+            return Arrays.stream(scoreDocs)
+                .map(scoreDoc -> findById(scoreDoc.doc + ":" + scoreDoc.score))
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .collect(toList());
+        } catch (IOException e) {
+            throw new RegistryIOException(e);
+        }
+    }
+
+    private ScoreDoc parseScoreDoc(String id) {
+        if (id == null) {
+            return null;
+        }
+        String[] components = id.split(":");
+        if (components.length != 2) {
```

Comment:
Could you, please, add some comment to this statement?

Reply:
Are you referring to the if statement? Added a comment. Please let me know if it's insufficient.

Reply:
Cool, thanks!